### ISSUE TITLE
Are documenttype drivers case-sensitive now on purpose?

### ISSUE THREAD

**Issue by sieversMartin** [plk/biblatex#946]
I was very much surprised today, when `biblatex` gave me the warning
```
Package biblatex Warning: No driver for entry type 'documenttype36_EXTEND_114'.
(biblatex)                Using fallback driver on input line 985.
```
The name of the driver is indeed `documenttype36_extend_114` while the entry type in the bib file is `@documenttype36_EXTEND_114`.

However, this has not been a problem at all with `biblatex` prior v3.14. Do I miss something here? Is this change intentional?

**Comment by moewew** [comment]
Yes, there was a change in how Biber handles capitalisation in the `.bib` file in Biber 2.14. See https://github.com/plk/biber/issues/278. The idea is that capitalisation is normalised to the capitalisation used in the data model declaration. Previously it was not well-defined or at least documented what would happen and apparently there was an inconsistency between field names and entry types.

I forgot to push the necessary documentation changes for `biblatex` 3.14 and only remembered to do that a few days ago (https://github.com/plk/biblatex/commit/ce59568b265b8bb2e079029c5caf8e9ee7081266) sorry about that. It was also not mentioned in CHANGES.md, which it should probably, so I'll change that in a sec.

**Comment by moewew** [comment]
Pushed https://github.com/plk/biblatex/commit/88793a487e7d8648ecaffbc85378711d309cb083 which mentions the changes in `CHANGES.md`.

Just to clarify: The intention of that change was to make sure that Biber/`biblatex` are **as case *in*sensitive as possible**. Of course the `biblatex`/LaTeX side of things is inherently case sensitive because TeX is, but BibTeX and Biber have always normalised `@ARTICLE` to `@article`. The change should make sure that all of this follows a logical patterns: Biber uses the capitalisation from the data model when it passes data on to `biblatex` (which is case sensitive). That means as long as all declarations and definitions on the `biblatex`/LaTeX side of things are consistent, things should work.

If that is not the case or you think the behaviour is otherwise undesirable, please consider posting a *full* example document that we can play around with.

**Comment by moewew** [comment]
@sieversMartin Did the comments (and commits) above make things clearer? I'd love to see an example of code that used to work, but doesn't any more to see if there is something that could be done.

**Comment by moewew** [comment]
Closing this for now. I would still be interested to see an MWE of what exactly we broke here. Most related issues that I have seen were somehow due to https://github.com/plk/biber/issues/299 and the fact that people didn't define their new entry types in a data model file (that is not the issue here as far as I can see).

If anyone has an MWE we can reopen this.

**Comment by sieversMartin** [comment]
Am 18.01.20 um 17:59 schrieb moewew:
> Closing this for now. I would still be interested to see an MWE of what
> exactly we broke here. Most related issues that I have seen were somehow
> due to plk/biber#299 <https://github.com/plk/biber/issues/299> and the
> fact that people didn't define their new entry types in a data model
> file (that is not the issue here as far as I can see).
> 
> If anyone has an MWE we can reopen this.

Sorry for my late response. Here is a MWE based on the original file,
which works well with biblatex v3.13 and breaks with v3.14 giving:

Package biblatex Warning: No driver for entry type
'documenttype36_EXTEND_114'.

(biblatex)                Using fallback driver on input line 43.

---
\begin{filecontents}{martin.bbx}
   \RequireBibliographyStyle{authortitle}

   \DeclareBibliographyDriver{documenttype36_extent_114}{%
      \usebibmacro{bibindex}%
      \usebibmacro{begentry}%
      \printfield{hbeschreibung}%
      \usebibmacro{finentry}%
   }%
\end{filecontents}

\begin{filecontents}{martin.cbx}
   \RequireCitationStyle{authortitle}
\end{filecontents}

\begin{filecontents}{martin.dbx}
   \DeclareDatamodelEntrytypes{%
      documenttype36_EXTEND_114,%
   }%

   \DeclareDatamodelFields[type=field,datatype=literal,nullok=true]{%
      hbeschreibung,%
   }%

   \DeclareDatamodelEntryfields[documenttype36_EXTEND_114]{%
      hbeschreibung%
   }

\end{filecontents}

\begin{filecontents}{martinexample.bib}
   @documenttype36_EXTEND_114cd{doc11841sub114_0,
      hbeschreibung = {Test},
}
\end{filecontents}
\documentclass{scrartcl}
\usepackage[T1]{fontenc}
\usepackage[style=martin]{biblatex}
\addbibresource{martinexample.bib}
\begin{document}
\nocite{*}

\printbibliography
\end{document}
---

Best regards,
Martin

**Comment by moewew** [comment]
There appear to be a few typos in the MWE, so I ran
```latex
\begin{filecontents}{martin.bbx}
\RequireBibliographyStyle{authortitle}

\DeclareBibliographyDriver{documenttype36_extend_114}{%
  \usebibmacro{bibindex}%
  \usebibmacro{begentry}%
  \printfield{hbeschreibung}%
  \usebibmacro{finentry}%
}%
\end{filecontents}

\begin{filecontents}{martin.cbx}
\RequireCitationStyle{authortitle}
\end{filecontents}

\begin{filecontents}{martin.dbx}
\DeclareDatamodelEntrytypes{%
  documenttype36_EXTEND_114,%
}%
\DeclareDatamodelFields[type=field,datatype=literal,nullok=true]{%
  hbeschreibung,%
}%
\DeclareDatamodelEntryfields[documenttype36_EXTEND_114]{%
  hbeschreibung%
}
\end{filecontents}

\begin{filecontents}{martinexample.bib}
@documenttype36_EXTEND_114{doc11841sub114_0,
  hbeschreibung = {Test},
}
\end{filecontents}
\documentclass{scrartcl}
\usepackage[T1]{fontenc}
\usepackage[style=martin]{biblatex}
\addbibresource{martinexample.bib}
\begin{document}
\nocite{*}

\printbibliography
\end{document}
```

It is a bit annoying that this doesn't work as before, but it is the effect of an intended change. Biber now normalises all entry types and field names to the variant used in the `.dbx` file (where the MWE has `documenttype36_EXTEND_114`) and not to all-lowercase as before (https://github.com/plk/biber/issues/278). That means that the entry type in the `.bbx`, `.cbx` and `.dbx` file should use the same uniform capitalisation.

This should not affect users who follow the convention of using all-lowercase names in their `.bbx`, `.cbx` and `.dbx` files.

Bottom line: Yes this is an expected change. It should be documented now. Sorry for the inconvenience, but I hope the new scheme is overall more logical.

**Comment by sieversMartin** [comment]
Sorry for the typo.

For me the change makes things more consistent for future applications. As I get the entrytypes from an database publishing system which uses mixedcase names, I always had to take care of the correct form in dbx, cbx, bbx and bib files.
